TFNetworkRecLayer.py

```diff
@@ -2352,35 +2352,38 @@ def cond(i, net_vars, acc_tas, seq_len_info=None):
 
     return output, output_search_choices
 
-  def _opt_search_resolve(self, layer_name, acc_ta, final_net_vars):
+  def _opt_search_resolve(self, layer_name, acc_ta, final_net_vars, seq_len):
     """
     This assumes that we have frame-wise accumulated outputs of the specific layer (acc_ta).
     If that layer depends on frame-wise search choices, i.e. if the batch dim includes a search beam,
     and each frame has different beams,
     then we will resolve that to the common final best search choices,
     which is determined by the latest search choice given by the layer.
     This assumes that we also have `self.final_acc_tas_dict["choice_%s" % choice_base.name]` available.
+    In addition, we resolve the sequence lengths (whose beams correspond to the end layer) to the final search choices.
 
     :param str layer_name:
     :param tf.TensorArray acc_ta: accumulated outputs of that layer
     :param final_net_vars:
-    :return: (new acc_ta, final search choices)
-    :rtype: (tf.TensorArray,SearchChoices|None)
+    :param tf.Tensor seq_len: shape (batch * beam,), has beam of the "end" layer in case of dynamic sequence lengths,
+      otherwise beam of rec_layer.output
+    :return: (new acc_ta, final search choices, resolved seq_len)
+    :rtype: (tf.TensorArray,SearchChoices|None,tf.Tensor)
     """
     import os
     from TFUtil import nd_indices, assert_min_tf_version, expand_dims_unbroadcast, get_valid_scope_name_from_str
     from TFUtil import get_shape_dim
-    from TFNetwork import LayerNotFound
     rec_layer = self.parent_rec_layer
     try:
       layer = self.net.get_layer(layer_name)
     except LayerNotFound:  # layer is not inside loop
-      return acc_ta, None
-    layer_choice = self.net.get_search_choices(src=layer)
-    if not layer_choice:
-      return acc_ta, None
-    if layer_choice.search_choices.keep_raw:
-      return acc_ta, layer_choice.search_choices
+      return acc_ta, None, seq_len
+    search_choices = layer.get_search_choices()
+    if not search_choices:
+      return acc_ta, None, seq_len
+    if search_choices.keep_raw:
+      return acc_ta, search_choices, seq_len
+    layer_choice = search_choices.owner
```

Comment:
Like this? I checked that search choices are the same as before for all test cases...

Reply:
Yes exactly. Yes I think I once had a case where this was more important, but changed that later on. But still, maybe this is better.